ginkgo/outline/outline.go

```diff
@@ -0,0 +1,102 @@
+package outline
+
+import (
+	"encoding/json"
+	"fmt"
+	"go/ast"
+	"strings"
+
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+const (
+	// ginkgoImportPath is the well-known ginkgo import path
+	ginkgoImportPath = "github.com/onsi/ginkgo"
+)
+
+// FromASTFile returns an outline for a Ginkgo test source file
+func FromASTFile(src *ast.File) (*outline, error) {
+	ginkgoImportName, ok := importNameForPackage(src, ginkgoImportPath)
+	if !ok {
+		return nil, fmt.Errorf("file does not import %s", ginkgoImportPath)
+	}
+	root := ginkgoNode{
+		Nodes: []*ginkgoNode{},
+	}
+	stack := []*ginkgoNode{&root}
+
+	ispr := inspector.New([]*ast.File{src})
+	ispr.Nodes([]ast.Node{(*ast.CallExpr)(nil)}, func(node ast.Node, push bool) bool {
+		if push {
+			// Pre-order traversal
+			ce, ok := node.(*ast.CallExpr)
+			if !ok {
+				// Because `Nodes` calls this function only when the node is an
+				// ast.CallExpr, this should never happen
+				panic(fmt.Errorf("node starting at %d, ending at %d is not an *ast.CallExpr", node.Pos(), node.End()))
```

Comment:
Not sure if panicking is the only option. This is _supposed_ to be impossible to reach.